examples/thread_pool.cpp

```diff
@@ -0,0 +1,48 @@
+#include "thread_pool.hpp"
+
+// Implementation:
+simple_thread_pool::simple_thread_pool(std::size_t sz)
+    : m_shared_src(std::make_shared<pool_src>()), size(sz)
+{
+    for (int i = 0; i < sz; ++i) {
+        std::thread(
+            [this](std::shared_ptr<pool_src> ptr) {
+                while (true) {
+                    std::function<void()> task;
+                    // >>> Critical region => Begin
+                    {
+                        std::unique_lock<std::mutex> lock(ptr->queue_mu);
+                        ptr->cv.wait(lock, [&] {
+                            return ptr->shutdown or !ptr->queue.empty();
+                        });
+                        if (ptr->shutdown and ptr->queue.empty())
+                            return;  // Conditions to let the thread go.
+                        task = std::move(ptr->queue.front());
+                        ptr->queue.pop();
+                    }
+                    // >>> Critical region => End
+                    task();
+                    if (ptr->to_finish.fetch_add(
+                            -1, std::memory_order_relaxed) == 1)
+                        ptr->wait_cv.notify_one();
+                }
+            },
+            m_shared_src)
+            .detach();
+    }
+}
+
+void simple_thread_pool::wait()
+{
+    std::unique_lock lock{m_shared_src->wait_mu};
+    m_shared_src->wait_cv.wait(lock, [&] {
+        return m_shared_src->to_finish.load(std::memory_order_relaxed) == 0;
+    });
+}
+
+simple_thread_pool::~simple_thread_pool()
+{
+    m_shared_src->shutdown = true;
```

Comment:
`m_shared_src->shutdown` is used in critical region of worker but not protected here.

Reply:
Only one thread will call the deconstructor :)

Reply:
But every worker is reading `m_shared_src->shutdown` in their loop:
https://github.com/tensorlayer/openpose-plus/blob/c24ef3ed37475f89a63e4385e83a99b2f34d70dd/examples/thread_pool.cpp#L15-L17

Reply:
Yes, it is. But this won't result in deadlock. The `shutdown` will only turn into `true` from `false` once, which `cv` is finally able to get to know when it's `true`.